File: notes/microsoft/lowest_common_ancestor.js
var lowestCommonAncestor = function (root, p, q) {
  if (root === null || root === p || root === q) {
    return root;
  }
  const left = lowestCommonAncestor(root.left, p, q);
  const right = lowestCommonAncestor(root.right, p, q);
  if (left && right) {
    return root;
  }
  if (left) {
    return left;
  } else {
    return right;
  }
};

// const lowestCommonAncestor = (root, p, q) => {
//   if (root === null || root === p || root === q) {
//     return root;
//   }
//   const left = lowestCommonAncestor(root.left, p, q);
//   const right = lowestCommonAncestor(root.right, p, q);
//   if (left && !right) { return left; }
//   if (right && !left) { return right; }
//   if (left && right) { return root; }
// };

File: 2022/lowest_common_ancestor.py
class Solution(object):
    def lowestCommonAncestor(self, root, p, q):
        return self.dfs(root, p, q, {'LCA': None})
        

    def dfs(self, root, p, q, map):
        if not root:
            return None
      
        left = self.dfs(root.left, p, q, map)
        right = self.dfs(root.right, p, q, map)

        if root.val == p.val or root.val == q.val:
            return root

        if left and right and not map['LCA']:
            map['LCA'] = root
            return map['LCA']
        if left:
            return left
        if right:
            return right


        return map['LCA']